Proposal for run bundle-upgrade (#27)
operator-framework/enhancements: enhancements/upgrade-bundle.md

---
title: upgrade-bundle
authors:
  - "@rashmigottipati"
reviewers:
  - TBD
approvers:
  - TBD
creation-date: 2020-06-11
last-updated: 2020-06-11

---

# Upgrade Bundle

## Release Signoff Checklist

- [ ] Enhancement is `implementable`
- [ ] Design details are appropriately documented from clear requirements
- [ ] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions

> 1. How do we handle multiple subscriptions? 

Reviewer: This question is critical to answer because this command needs smart detection logic for whether we need to create a new `Subscription` (same channel? same starting CSV? same operator?) or error out.

@njhale @ecordell are `Subscription`'s patch-able, ex. can the starting CSV or channel be updated?

> 1. How do we handle multiple subscriptions? 

Author: I don't have a specific use-case for multiple subscriptions. Maybe @estroz does.

> 1. How do we handle multiple subscriptions? 

Author: I updated the PR to include these details for that open question.

## Open Questions

1. How do we handle multiple subscriptions? 
2. Should we explicitly pass new version number as a flag or can we get it from bundle image?

>     * Based on discussions, can extract the version number from bundle image. 

Reviewer: :+1:

> 3. How do we verify that the upgrade is successful 

Reviewer: For now let's just check if the CSV has status `Installed` in the correct set of namespaces. Eventually OLM will/should surface the result of a deeper inspection of a successful upgrade (ex. pod readiness) somewhere, like in a CSV's status perhaps, but that's clearly out of scope for this EP.

Author: Thanks for throwing some light on these open questions 👍  , I've incorporated your feedback into those.

> 4. Is it valid to assume that no previous version of the operator is already running? 

Reviewer: I think the scope of this ep should be narrowed to "implement `run bundle-upgrade` for operators that don't exist in a cluster" since the "existing operator" scenario is different implementation-wise.

Enhance the Operator SDK to add a new subcommand `bundle-upgrade` that performs an operator upgrade using OLM.

## Motivation

Testing operator upgrades remains difficult today due to the number of steps to prepare for an upgrade and the lack of tooling that automates those steps. Having high quality Operators is crucial to the success of the Operator Framework and thereby operator developers should have a tool to use to perform an operator upgrade.

### Goals

Provide a command line that upgrades an operator version using OLM.

### Non-Goals

> 

Author: Totally agree! Added a few non-goals based on some assumptions we are making. Please do suggest if you think we need to add anything else to those.

### Non-Goals

## Proposal

The Operator SDK currently supports running an operator with OLM using the

> `run` command with the `--olm` flag. 

Reviewer: The `--olm` flag is deprecated.
```suggestion
`run packagemanifests` subcommand
```

Currently, the run command has different subcommands: 

 * `run --local` - exists today; converting to a subcommand will be covered by [OSDK-1126](https://issues.redhat.com/browse/OSDK-1126)

>  * `run packagemanifest` - covered by [OSDK-1126](https://issues.redhat.com/browse/OSDK-1126)

Author: Got it, updated

 * `run bundle` - currently being implemented, design document is covered by [run bundle enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/run-bundle.md)

#### Command Line Interface
The new command line will look as follows:

```
operator-sdk run bundle-upgrade (alias: bu) <bundle-image> --from-index=<> --starting-csv=<> [--channel=<>] [--operator-namespace=<>] [--install-mode=<>]
```

* Positional arguments:
    * `bundle-image`
* Required flags:

>     * `--from-index` 

Author: Good idea. Renamed it

    * `bundle-image`
* Required flags:
    * `--index-image` 
    * `--starting-csv`
* Optional flags:
    * `--channel` 
    * `--operator-namespace`
    * `--install-mode`

* `<bundle-image>` is a positional argument that specifies the bundle image that we want to upgrade to
* `--index-image=<>` is a required flag, which contains the operator we’re upgrading from

> * `--starting-csv=<>` is a required flag, which specifies the first version of the CSV that OLM will install as part of the subscription for a given package. 

Reviewer: There's a lot of information packed into these flag descriptions. Can you write a paragraph or two about how you'll use these values in concert in an implementation details section?

Author: From my understanding of this comment, it's to handle scenarios of simple (requiring minimal flags)/complex (providing all the flags) example usages.
I just expanded on it in the `Example UX scenarios` section. Let me know if this is what you were expecting. If not, happy to discuss further.

> * `--starting-csv=<>` is a required flag

Author: Added a link to a sample CSV of memcached operator from operator-sdk-samples

>   the operator. It will default to the `KUBECONFIG` context.

Author: Good catch! As you said, KUBECONFIG context has no impact on channel. I intended to have that line for operator namespace and added it for channel description too by mistake. Removed it

* `[--channel=]` is an optional flag that specifies the channel to upgrade from within a subscription (default channel used if subscription does not contain a channel)
* `[--operator-namespace=]` is an optional flag that specifies the namespace to install the operator. It will default to the `KUBECONFIG` context. (see run bundle)
* `[--install-mode=]` is an optional flag that specifies the
  [OLM install mode](https://red.ht/2YK0ksD). (see run bundle)
  * We support 3 modes:
    * AllNamespaces
    * OwnNamespace
    * SingleNamespace=ns1
      * NOTE: if ns1 is the same as the operator's namespace, a warning will be
        printed
  * When `--install-mode` is not supplied, the default precedence order is:

>     * `OperatorGroup`, if it already exists

Reviewer: This should be more like:
```markdown
    * If an `OperatorGroup` already exists with the correct namespacing, use it.
    * Else, create an `OperatorGroup` with the following preferences:
    	* If `AllNamespaces` is supported, use this mode.
    	* Else if `OwnNamespace` is supported, use this mode.
    	* Else:
			* If `SingleNamespace` is not supported, return an error saying "no valid install mode is supported"
			* Else return an error saying "--install-mode=SingleNamespace=<namespace> is required"
```

Also move this to an implementation details section.

Author: Modified this as per your suggestion. Will move this to implementation details as well

    * `OwnNamespace`, if the CSV supports it
    * If the CSV does not support either of these, `--install-mode` flag
      will be **required**

#### Assumptions
* OLM is already installed
* No previous version of the operator is already running
* An index image exists that has the from version of the operator bundle

> * New version of the operator exists in the cluster 

Reviewer 2: What does this mean? Is it just that the `<bundle-image>` and any images it references are pullable by the cluster? Just want to make sure we're precise, because I could see this being interpreted to mean that the new version of the operator is already installed and running, which would be kinda weird :)

Author: Absolutely. 
I meant that the provided images should be accessible and pullable in the cluster by kubelet, I rephrased it

* The provided images (i.e. bundle and index) are accessible and pullable by kubelet 

#### Example UX Scenario

```

> operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv csv.v1 --channel alpha

Reviewer: Nit: more realistic naming
```suggestion
operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --channel alpha
```

 1. Create pod using `--from-index`, set container command (entrypoint) to:
    ```
    /bin/bash -c ‘ \
      /bin/mkdir -p /database && \
      /bin/opm registry add   -d /database/index.db -b {.BundleImage} && \
      /bin/opm registry serve -d /database/index.db’
    ```
 1. If pod fails, clean it up, and error out with message of unsupported index image. Also capture the pod's logs for debugging.

>  1. Create GRPC CatalogSource, pointing to <pod-name>.<namespace>:50051

Reviewer 2: Markdown gets confused on the angle brackets. I think enclosing in backticks should fix it.

```suggestion
 1. Create GRPC CatalogSource, pointing to `<pod-name>.<namespace>:50051`
```

 1. Create OperatorGroup, only if it does not exist
 1. Create Subscription

>     * Use `--channel` as alpha, `--starting-csv` as csv.v1, and the operator name from the provided bundle.

Reviewer 2: Minor clarity improvement:
```suggestion
    * Use `--channel` as "alpha" and `--starting-csv` as "csv.v1", taken from the flag values.
    * Use the operator name from the provided bundle.
```

Author: Done

>     * Use `--channel` as "alpha" and `--starting-csv` as "csv.v1", taken from the flag values.

Reviewer: ```suggestion
    * Use values from `--channel` and `--starting-csv`, and the created `CatalogSource`'s name.
```

Author: Good catch! Realized that we need CS' name in order to create a subscription, fixed it

    * Configure approval to automatic upgrades

>  1. Operator installed into the default namespace defined by KUBECONFIG

Reviewer: ```suggestion
1. Operator installed into the default namespace defined by KUBECONFIG
```

1. Verify operator is upgraded (by looking at the CSV version, status, etc.)

> 1. Log what happened and final state (e.g. operator is running in namespace X, watching namespace(s) Y.

Reviewer: ```suggestion
1. Log what happened and final state (e.g. operator is running in namespace X, watching namespace(s) Y).
```

> 1. Log command to execute to cleanup operator.

Reviewer: What happens if an error occurs in one of the above steps? Do any resources get cleaned up?

Discuss this scenario in an implementation details section.

```

* Positional arguments:
    * `bundle-image`
* Required flags:
    * `--index-image` 
    * `--starting-csv`
* Optional flags:
    * `--channel` 
    * `--operator-namespace`
    * `--install-mode`

* `<bundle-image>` is a positional argument that specifies the bundle image that we want to upgrade to
* `--index-image=<>` is a required flag, which contains the operator we’re upgrading from
* `--starting-csv=<>` is a required flag, which specifies the first version of the CSV that OLM will install as part of the subscription for a given package. 
Here's an [sample CSV](https://github.com/operator-framework/operator-sdk-samples/blob/master/go/memcached-operator/deploy/olm-catalog/memcached-operator/manifests/memcached-operator.clusterserviceversion.yaml) in which the value of startingCSV is the name i.e. `memcached-operator.v0.17.0`
* `[--channel=]` is an optional flag that specifies the channel to upgrade from within a subscription (default channel used if subscription does not contain a channel)
* `[--operator-namespace=]` is an optional flag that specifies the namespace to upgrade the operator. It will default to the `KUBECONFIG` context. (see run bundle)

> * `[--install-mode=]` is an optional flag that specifies the

Author: Explained in Example UX Scenarios section

  [OLM install mode](https://red.ht/2YK0ksD). (see run bundle)
  * We support 3 modes:
    * AllNamespaces
    * OwnNamespace
    * SingleNamespace=ns1

>       * NOTE: if ns1 is the same as the operator's namespace, a warning will be

Reviewer 3: @shawn-hurley this is exactly how we worded it for run bundle. I suspect it would be just a log message to the user stating they can't be the same.

  * Behavior of `run bundle-upgrade` in the presence of an OperatorGroup
    * If an operator is installed, it should have a Subscription and an OperatorGroup already created. This OperatorGroup exists with the correct namespacing and it satifies the subscription. So `run bundle-upgrade` will use the same OperatorGroup that was created during install. However, if the upgrade of the operator doesn't work with the existing OperatorGroup, then error out. 
    In `run bundle`, there is support for 3 modes: 
        * AllNamespaces
        * OwnNamespace
        * SingleNamespace=ns1

Example UX Scenarios are discussed below. 

#### Command Line Interface
The new command line will look as follows:

```
operator-sdk run bundle-upgrade (alias: bu) <bundle-image> --index-image=<> --subscription-name=<> --namespace=<> [--mode=<>]
```

* Positional arguments:
    * `bundle-image`
* Required flags:
    * `--index-image` 

>     * `--subscription-name`

Reviewer: Since this is required, we should have `run bundle` print this name after returning @jmrodri.

Reviewer 3: @estroz that makes sense

* Optional flags:
    * `--mode` 
    * `--namespace`

* `<bundle-image>` is a positional argument that specifies the bundle image that we want to upgrade to.
* `--index-image=<>` is a required flag, which contains the operator we’re upgrading from.
* `--subscription-name=<>` is a required flag that specifies the name of a subscription. 

> * `[--mode=]` is an optional flag that specifies the graph update mode that defines how channel graphs are updated. Mode can be one of: [replaces, semver, semver-skippatch] (default "replaces"). It will default to the `replaces`.

Reviewer 4: I thought that we talked about a hidden flag here?

* `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --channel alpha --namespace foo --install-mode SingleNamespace=bar`

#### Simple Example

Here's a simple example with all the required flags.

```
operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0
```

Below is the workflow when the above command is invoked:

 1. Create pod using `--index-image`, set container command (entrypoint) to:
    ```
    /bin/bash -c ‘ \
      /bin/mkdir -p /database && \
      /bin/opm registry add   -d /database/index.db -b {.BundleImage} && \
      /bin/opm registry serve -d /database/index.db’
    ```
 1. If pod fails, clean it up, and error out with message of unsupported index image. Also capture the pod's logs for debugging.
 1. Create GRPC CatalogSource, pointing to `<pod-name>.<namespace>:50051`
 1. If an OperatorGroup doesn't exist, create one and default to `AllNamespaces` as `--install-mode` was not provided
 1. Create Subscription

>     * Use default values for `--channel` and `--starting-csv`, and the created `CatalogSource`'s name.

Reviewer 4: How did we determine the default for the `starting-csv` I think we mention above that we pull it from the bundle, can we detail the steps here if we are going to mention it?

Author: `starting-csv` is required, so it has to be specified. I meant to use default value for channel alone. Rephrased it

 1. Create GRPC CatalogSource, pointing to `<pod-name>.<namespace>:50051`
 1. Creates an OperatorGroup in and targetting namespace `default`.
 1. Create Subscription
    * Use values from `--channel` and `--starting-csv`, and the created `CatalogSource`'s name.
    * Configure approval to automatic upgrades
    
 1. Operator upgraded in the namespace `memcached-system`
 1. Verify operator is upgraded (by looking at the CSV version, status, etc.)
 1. Log what happened and final state (e.g. operator is running in namespace
    `memcached-system`, watching namespace `default`.
 1. Log command to execute to cleanup operator.
Exit with success (leaving operator running)

#### Prototype Links

 * Prototype for run bundle may be able to be adapted to prototype run bundle-upgrade: https://github.com/joelanford/bundle-runner
 * Bundle format:
   https://github.com/operator-framework/operator-registry/blob/master/docs/design/operator-bundle.md

### Test Plan

  * manual testing with different scenarios 
  * add unit tests for `internal/olm/operator/operator.go` (with coverage requirements)

>   * investigate adding e2e tests using OLM

Reviewer 4: I would like to see this be more of a statement of we will add end to end tests. Are we not e2e testing the olm cli stuff yet?

* `[--channel=]` is an optional flag that specifies the channel to upgrade from within a subscription (default channel used if subscription does not contain a channel)
* `[--operator-namespace=]` is an optional flag that specifies the namespace to upgrade the operator. It will default to the `KUBECONFIG` context. (see run bundle)
* `[--install-mode=]` is an optional flag that specifies the
  [OLM install mode](https://red.ht/2YK0ksD). (see run bundle)
  * We support 3 modes:
    * AllNamespaces
    * OwnNamespace
    * SingleNamespace=ns1
      * NOTE: if ns1 is the same as the operator's namespace, a warning that "ns1 can not be the same as the operator's namespace" will be printed
  * When `--install-mode` is not supplied, the selection algorithm is as follows:
    * If an `OperatorGroup` already exists with the correct namespacing, use it.
    * Else, create an `OperatorGroup` with the following preferences:
    	* If `AllNamespaces` is supported, use this mode.
    	* Else if `OwnNamespace` is supported, use this mode.
    	* Else:
			* If `SingleNamespace` is not supported, return an error saying "no valid install mode is supported"

> 			* Else return an error saying "--install-mode=SingleNamespace=<namespace> is required"

Reviewer: Can you move this section to implementation details and keep the simple flag explanation?

Author: Good idea, moved it.

Let's consider a few scenarios: 

How do we verfiy whether an operator has been upgraded successfully?
To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully installed or upgraded. 
  
In case of subscription failure, the end user needs to be aware of this. If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. 

> We then log a message that "run bundle-upgrade was not successful", clean u the registry pod and exit.

Reviewer 4: It might be useful to keep the registry pod there to maybe determine the graph that was being used. I think that flag to keep this on failure might be helpful?

Author: Yes, great point! I don't see any reason why the registry pod alone needs to be cleaned up. We could use a --skip-cleanup flag like scorecard to offer flexibility to operator developers for clean up options. 
Updated it.

---
title: bundle-upgrade
authors:
  - "@rashmigottipati"
reviewers:
  - "@joelanford"
  - "@estroz"
  - "@shawn-hurley"
  - "@jmrodri"
approvers:
  - TBD
creation-date: 2020-06-11
last-updated: 2020-06-11

---

# Bundle Upgrade

## Release Signoff Checklist

- [X] Enhancement is `implementable`
- [X] Design details are appropriately documented from clear requirements
- [X] Test plan is defined
- [ ] Graduation criteria for dev preview, tech preview, GA

## Open Questions

1. How do we handle multiple subscriptions? Are subscriptions patch-able ex. can the starting CSV or channel be updated for the same subscription?
    * Channel can be updated, starting CSV will not trigger an update if already installed.
2. Should we explicitly pass new version number as a flag or can we get it from bundle image?

>     * Based on discussions, can extract the version number from bundle image. 

Reviewer 4: correct, we said that we would use the default `replace` mode for the case of adding to an index that exists but semver for an empty index

I think the same applies here IMO, that we use the default `replace` mode for now.

In case 1, we will have to figure out which catalog source the existing operator was installed in and somehow dynamically inject the new bundle image into the index image for OLM to detect this change and upgrade to the new version we provided. 

In this proposal, we will design bundle upgrade for case 2. For this case, we also assume that an index image already exists that has the from version of the operator bundle. So we don't create a new index image. We use the existing index image (which is mandatory to provide) for the operator and provide the new bundle image for the upgrade. For this to happen, we need to specify starting CSV.
Generally, when there is no starting CSV specified, OLM always detects and installs the latest version available for an operator in a particular  channel. For example, in production case, if an operator is running and there's a subscription for it, and there comes a fix available in a stable channel, OLM will detect and upgrade by default. 

Let's consider a few scenarios: 

> How do we verfiy whether an operator has been upgraded successfully?

Reviewer 2: typo

```suggestion
How do we verify whether an operator has been upgraded successfully?
```

To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully installed or upgraded. 
  

> In case of subscription failure, the end user needs to be aware of this. If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. 

Reviewer 2: This gets back to my question above about whether we need to know the version we're upgrading to. For us to verify that the upgrade ends with the bundle provided on the command line, I agree that this seems like a good approach.

There is one possible gotcha though: how do we know if OLM is done upgrading? Here's a scenario:
* Starting CSV: `A`
* Bundle provided on command line: `B`
* Upgrade graph of existing index image: `A -> C`
* Upgrade graph of ephemeral index after inserting B: `A -> B -> C`

After OLM upgrades A to B, our check would say the upgrade was successful, but then simultaneously OLM is probably gearing up for the upgrade from `B` to `C`. Our command then returns saying we successfully upgraded, but then the user goes and looks at his cluster and sees that `C` is installed when they expected `B`.

Technically, OLM is never "finished" making upgrades, right? The entire idea is that if a new operator shows up in the upgrade graph, OLM will upgrade it.

@shawn-hurley @ecordell For our purposes, is there a reasonable way to know when OLM has made its way though the upgrade graph of the channel in our ephemeral index?

Another similar scenario is:
* Starting CSV: `A`
* Bundle provided on command line: `Z`
* Upgrade graph of existing index image: `A -> B -> C -> ... -> X -> Y`
* Upgrade graph of ephemeral index after inserting Z: `A -> B -> C -> ... -> X -> Y -> Z`

This would take OLM awhile to roll through right?

Is there a way for SDK to know the upgrade graph for a channel starting with `--starting-csv` and ending with the HEAD of the channel?

Reviewer: Scenario 1: all upgrades up to `B` were successful, and that's all the command guarantees to report right? If `run bundle-upgrade` tries to report on operator versions "later" in the upgrade graph then semantically the command is always checking the entire upgrade channel and not one specific upgrade path.

Scenario 2: the naive approach would be to poll for operator version `Z` until a timeout ends or `Z` installs/fails to install. One way I see around having to tune the timeout based on how many operators are in an upgrade graph is, like you suggest, to rebuild the graph and poll for each operator in it until we reach `Z`, resetting the timeout for each correct installation. I'm sure there's a way to get this info from the index db.

2. Should we explicitly pass new version number as a flag or can we get it from bundle image?
    * Based on discussions, can extract the version number from bundle image. 

## Summary

Enhance the Operator SDK to add a new subcommand `run bundle-upgrade` that performs an operator upgrade using OLM.

## Motivation

Testing operator upgrades remains difficult today due to the number of steps to prepare for an upgrade and the lack of tooling that automates those steps. Having high quality Operators is crucial to the success of the Operator Framework and thereby operator developers should have a tool to use to perform an operator upgrade.

### Goals

Provide a command line that upgrades an operator version using OLM.

### Non-Goals

This enhancement doesn't include the capability for SDK to use labels to detect the DB path. For the sake of simplicity, we will assume that the existing registry DB is located at `/database/index.db`.

> Another non-goal is that the enhancement is intended for operators that don't exist in the cluster. The scenario in which existing operators exist in the cluster is out of scope for this proposal.

Reviewer 5: If this feature is limited like this, then we can not test that operands will be upgraded by the new version of an operator.

Author: Thanks for your input Matskiv. Updated the PR with revised proposal which can upgrade from a previously installed operator to a new specified version. Please take a look.

Reviewer 5: Thank you for addressing use case for my team.
I skimmed through the proposal and it looks good to me, but your team is, of course, in a better position to properly asses technical side of the proposal. :)

Author: thank you so much for your valuable feedback and explaining your use case in our offline discussion. glad that this will feature will be useful for your use case!

## Proposal

### Implementation Details/Notes/Constraints [optional]

The Operator SDK currently supports running an operator with OLM using the
`run packagemanifests` subcommand. 
This enhancement will add a new subcommand, `bundle-upgrade`,
to the `run` command. This new command, `run bundle-upgrade`, will perform the upgrade of an operator from a previous version to a new specified version.

Currently, the run command has different subcommands: 

 * `run local`
 * `run packagemanifests`
 * `run bundle` - currently being implemented, design document is covered by [run bundle enhancement](https://github.com/operator-framework/enhancements/blob/master/enhancements/run-bundle.md)

For doing an `bundle-upgrade`, there could be two scenarios: 

* Case 1 - A previous version of the operator is already running on the cluster, and the index image contains this operator version. Now, we we want to upgrade the in cluster running operator to a new version that we specify. 

> * Case 2 - No previous version of an operator is running on the cluster, so we provide the new bundle image which is accessible and pullable within the cluster, which contains our desired version of the operator for the upgrade. 

Reviewer 6: The way this is described, isn't it just install? or is the case of installing an initial version, and then upgrading to the target?

Author: This isn't just install. This command takes `--starting-csv` as a required flag, which specifies the first version of the CSV that OLM will install, and a positional argument `<bundle-image>` that specifies the newer version of the operator that we want to upgrade to.

> In case 1, we will have to figure out which catalog source the existing operator was installed in and somehow dynamically inject the new bundle image into the index image for OLM to detect this change and upgrade to the new version we provided. 

Reviewer 6: Actually, OLM is capable of resolving over multiple catalogs. IIRC, you can just create a new index that contains a replacement bundle in the same package/channel as the current `Subscription` and OLM will perform the upgrade for you (should the `Subscription` be on "automatic approval").

Author: This seems totally fine. But in production, can we create a new index while the previous version of the operator (with the old index) is already running in the cluster and dynamically modify the image reference to point to this new index image we just built? Is that possible?

Author: If we set the approval to `automatic`, OLM always updates to the latest version available in the channel we specify in this command, in which case testing the operator version we wanted to upgrade to, might be difficult. Thoughts?

In this proposal, we will design bundle upgrade for case 2. For this case, we also assume that an index image already exists that has the from version of the operator bundle. So we don't create a new index image. We use the existing index image (which is mandatory to provide) for the operator and provide the new bundle image for the upgrade. For this to happen, we need to specify starting CSV.
Generally, when there is no starting CSV specified, OLM always detects and installs the latest version available for an operator in a particular  channel. For example, in production case, if an operator is running and there's a subscription for it, and there comes a fix available in a stable channel, OLM will detect and upgrade by default. 

Let's consider a few scenarios: 

  * Verifying operator upgrades

>     * To verify whether an operator upgrade to a desired version has been successful, the ClusterServiceVersion (CSV) CustomResource is a useful resource for providing such context to the end user. A given subscription has its status field and corresponding InstallPlan. If a ClusterServiceVersion (CSV) is created for that InstallPlan, then the CSV Status indicates that the operator has been successfully upgraded. We will ensure that the CSV has status `Installed` in the appropriate namespaces.  

Reviewer 6: I think you want to make sure the CSV status is `Successful` rather than `Installed` and that the previous CSV has been deleted.

Author: Fixed it.

  * Handling subscription failures 

>     * If a subscription failed, there might not be an InstallPlan that's created for the new operator bundle. However, through the SubscriptionStatus, we can obtain the CurrentCSV and compare it with the desired operator version we wanted to upgrade to, and verify that they are not the same, and hence conclude there's been a failure. We then log a message that "run bundle-upgrade was not successful", and exit. 

Reviewer 6: What happens if there are multiple versions in the chain between the InstalledCSV and CurrentCSV? OLM's behavior dictates that we we step through each upgrade. The state described here is transient while upgrades are progressing, so I wouldn't outright fail the upgrade based on this condition.

  * Operator fails to come up after creating subscription 
    * If operator fails to come up after creating the subscription, it failed to create a second install plan and the SubscriptionStatus will indicate that. Therefore, we proceed to logging the error message and exiting. 

  * What happens when a catalog source is not created?
    * When a catalog source fails to create, it would show on the status of the CatalogSource. We log an error message, and exit. 

  * Do we cleanup resources after failures or not? 
    * We could offer flexibility to the operator developers to use a flag --skip-cleanup to indicate whether or not cleanup of resources is desired. This could enable the operator developers to capture all the logs and inspect the pod failures.

  * Behavior of `run bundle-upgrade` in the presence of an OperatorGroup

>     * It is important to note that the presence of an OperatorGroup in the namespace can affect the how `run bundle-upgrade` operates. The presence of the OperatorGroup will take precedence. If the OperatorGroup does not support  version an error condition will occur. The `run bundle-upgrade` command will return with a non-zero exit status and log the errors. 

Reviewer 6: ```suggestion
    * It is important to note that the presence of an OperatorGroup in the install namespace can affect the operation of `run bundle-upgrade`. An existing OperatorGroup will take precedence. If the OperatorGroup does not support the InstallMode of the operator, the `run bundle-upgrade` command will return with a non-zero exit status and log an error. 
```

  * Install Mode and OperatorGroups
    * We support 3 modes:
      * AllNamespaces
      * OwnNamespace
      * SingleNamespace=ns1
        * NOTE: if ns1 is the same as the operator's namespace, a warning that "ns1 can not be the same as the operator's namespace" will be printed
    * When `--install-mode` is not supplied, the selection algorithm is as follows:
      * If an `OperatorGroup` already exists with the correct namespacing, use it.

>       * Else, create an `OperatorGroup` with the following preferences:

Reviewer 6: There's a missing case here: The `OperatorGroup` exists and doesn't match the given `--install-mode`.

Reviewer: In this case `bundle-upgrade` should error out.

    	  * If `AllNamespaces` is supported, use this mode.
    	  * Else if `OwnNamespace` is supported, use this mode.
    	  * Else:
			  * If `SingleNamespace` is not supported, return an error saying "no valid install mode is supported"
			  * Else return an error saying "--install-mode=SingleNamespace=<namespace> is required"

Example UX Scenarios are discussed below. 

#### Command Line Interface
The new command line will look as follows:

```
operator-sdk run bundle-upgrade (alias: bu) <bundle-image> --index-image=<> --starting-csv=<> [--channel=<>] [--operator-namespace=<>] [--install-mode=<>]
```

* Positional arguments:
    * `bundle-image`
* Required flags:
    * `--index-image` 
    * `--starting-csv`
* Optional flags:
    * `--channel` 
    * `--operator-namespace`
    * `--install-mode`

* `<bundle-image>` is a positional argument that specifies the bundle image that we want to upgrade to
* `--index-image=<>` is a required flag, which contains the operator we’re upgrading from

> * `--starting-csv=<>` is a required flag, which specifies the first version of the CSV that OLM will install as part of the subscription for a given package. 

Reviewer 6: nit: since these are required, should the be positional parameters instead of arguments?

Here's an [sample CSV](https://github.com/operator-framework/operator-sdk-samples/blob/master/go/memcached-operator/deploy/olm-catalog/memcached-operator/manifests/memcached-operator.clusterserviceversion.yaml) in which the value of startingCSV is the name i.e. `memcached-operator.v0.17.0`
* `[--channel=]` is an optional flag that specifies the channel to upgrade from within a subscription (default channel used if subscription does not contain a channel)
* `[--operator-namespace=]` is an optional flag that specifies the namespace to upgrade the operator. It will default to the `KUBECONFIG` context. (see run bundle)
* `[--install-mode=]` is an optional flag that specifies the
  [OLM install mode](https://red.ht/2YK0ksD). (see run bundle)

#### Assumptions
* OLM is already installed
* No previous version of the operator is already running
* An index image exists that has the from version of the operator bundle

> * The provided images (i.e. bundle and index) are accessible and pullable by kubelet 

Reviewer 6: If this is to be used in the operator dev process, in-series with bundle building, I don't think it's too far of a stretch to say the user has push access to a cluster-accessible registry. In which case building/pushing index images isn't too far fetched. Thoughts?

Reviewer: Are you suggesting this command optionally builds/pushes an index image instead of pulling one?

* Given an operator that supports install modes `AllNamespaces`, `OwnNamespace`, and `SingleNamespace`, being upgraded into a namespace with an existing OperatorGroup that targets all namespaces:
  * `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0` 
    * Upgrade into the default namespace configured in the `KUBECONFIG` context. The operator would watch all namespaces.
  * `operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0 --namespace memcached-system --install-mode SingleNamespace=default`
    * Upgrade into namespace `memcached-system`. The operator would watch namespace `default`.

#### Simple Example

Here's a simple example with all the required flags.

```
operator-sdk run bundle-upgrade <bundle-image> --index-image quay.io/operator-framework/upstream-opm-builder --starting-csv test-operator.v1.0.0
```

Below is the workflow when the above command is invoked:

 1. Create pod using `--index-image`, set container command (entrypoint) to:
    ```

>     /bin/bash -c ‘ \

Reviewer 6: this assumes the index contains `bash`, which may not always be the case.

Author: this is one of the assumptions in this proposal that the index image has bash to be able to change the entrypoint. 
Will explicitly state this as an assumption

      /bin/mkdir -p /database && \
      /bin/opm registry add   -d /database/index.db -b {.BundleImage} && \
      /bin/opm registry serve -d /database/index.db’
    ```
 1. If pod fails, clean it up, and error out with message of unsupported index image. Also capture the pod's logs for debugging.
 1. Create GRPC CatalogSource, pointing to `<pod-name>.<namespace>:50051`
 1. If an OperatorGroup doesn't exist, create one and default to `AllNamespaces` as `--install-mode` was not provided

>  1. Create Subscription

Reviewer 6: nit: should these list numbers be increasing?

Author: in markdown, this will render a monotonically increasing sequence in the output automatically. The reason this is also allowed because when we make list additions, we don't have to keep tweaking the numbers. :)
Although I agree that it's confusing when we don't know about this feature
